### PR TITLE
Geofencing: Fetch and persist fences

### PR DESCRIPTION
**Pull request by evan-masseau**
# Description

Use the new API request (added in #342) to fetch and save geofence data from the API. Up next: monitor them!

## Due Diligence

- [ ] I have tested this on an emulator and/or a physical device.
- [x] I have added sufficient unit/integration tests of my changes.
- [ ] I have adjusted or added new test cases to team test docs, if applicable.
- [ ] I am confident these changes are compatible with all Android versions the SDK currently supports.

## Release/Versioning Considerations

- [ ] `Patch` Contains internal changes or backwards-compatible bug fixes.
- [ ] `Minor` Contains changes to the public API. 
- [ ] `Major` Contains **breaking** changes.
- [ ] Contains readme or migration guide changes.
  - If so, please merge to a feature branch so documentation updates only go live upon official release.
- [ ] This is planned work for an upcoming release.
  - If no, author or reviewer should account for this in a release plan, or describe why not below.

## Changelog / Code Overview

- Converted the location manager from object to class, use registry to keep it singleton (now that it is stateful, feels more appropriate to be a class)
- Fleshed out the `fetch` method to actually call the API
- Hooked up to persistent store so we save geofences to disk now
- Added `onGeofencesSynced` observer -- will be used as trigger to update fences monitored by the system, as well as a hook for test app to display geofences in map UI

## Test Plan

Unit tests for now, sample app integration is 1-2 PRs away and then we can fully test

## Related Issues/Tickets
 
https://klaviyo.atlassian.net/browse/CHNL-25306
https://klaviyo.atlassian.net/browse/CHNL-24475
 

---

> [!NOTE]
> Implements geofence fetch/persist with observer callbacks, converts the location manager to a class with registry accessor, fixes unregister behavior, updates receiver, and adds comprehensive tests; bumps API revision.
> 
> - **Location/Geofencing**:
>   - Convert `KlaviyoLocationManager` from `object` to `class` and add `Registry.locationManager` accessor.
>   - Implement fetching via `ApiClient.fetchGeofences`, persistence (`storeGeofences`/`clearStoredGeofences`/`getStoredGeofences`), and observer API (`onGeofenceSync`/`offGeofenceSync`); monitor stored geofences and use `NEVER_EXPIRE`.
>   - Fix `Klaviyo.unregisterGeofencing()` to call `stopGeofenceMonitoring`; instantiate manager in `registerGeofencing()`.
>   - Update `KlaviyoGeofenceReceiver` to route to `Registry.locationManager.handleGeofenceIntent`.
>   - Add `JSONArray.toKlaviyoGeofences` helper.
>   - Replace `LocationManager` interface with new contract (observers, storage, `handleGeofenceIntent`); remove old file.
> - **Tests**:
>   - Add `KlaviyoLocationManagerTest` covering fetch, persistence, observers, and storage edge cases.
>   - Update `GeofencingTest` to verify `stopGeofenceMonitoring`.
> - **Config**:
>   - Bump `klaviyoApiRevision` to `2025-10-15.pre`.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 2d3b48814a67313ecd19d4091bccf42577ad2ef1. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: sdk/location/src/main/java/com/klaviyo/location/Geofencing.kt [hunk 1]
```diff
@@ -27,7 +27,7 @@ fun Klaviyo.registerGeofencing(): Klaviyo = safeApply {
  * Klaviyo will automatically stop monitoring if location permissions are revoked.
  */
 fun Klaviyo.unregisterGeofencing(): Klaviyo = safeApply {
-    Registry.getOrNull<LocationManager>()?.startGeofenceMonitoring() ?: run {
+    Registry.getOrNull<LocationManager>()?.stopGeofenceMonitoring() ?: run {
```
Comment (evan-masseau) [inline sdk/location/src/main/java/com/klaviyo/location/Geofencing.kt]: whoops!

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 2]
```diff
@@ -7,27 +7,55 @@ import android.content.Context
...
+import com.klaviyo.analytics.networking.requests.FetchGeofencesResult
 import com.klaviyo.core.Registry
+import java.util.Collections
+import java.util.concurrent.CopyOnWriteArrayList
+import org.json.JSONArray
+
+/**
+ * Extension property to access a shared [LocationManager] instance
+ */
+internal val Registry.locationManager: LocationManager
```
Comment (evan-masseau) [inline sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: Because a geofence event intent could cold-start the app, we can't rely on the application code having already run `Klaviyo.registerGeofences` so we need a fallback for creating this service.

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 3]
```diff
@@ -61,6 +89,20 @@ internal object KlaviyoLocationManager : LocationManager {
         }
     }
 
+    override fun onGeofenceSync(callback: GeofenceObserver) {
```
Comment (evan-masseau) [inline sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: This observer will be the basis for
- updating monitored geofences with the system location APIs
- test app UI to display current fences

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 4]
```diff
@@ -73,17 +115,106 @@ internal object KlaviyoLocationManager : LocationManager {
         }
     }
```
[the remark this answers is not in the document]
Comment (evan-masseau) [reply sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: Yes, will deal with that later.

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 5]
```diff
@@ -73,17 +115,106 @@ internal object KlaviyoLocationManager : LocationManager {
...
+            }
+        }
+    }
+
+    /**
+     * Save geofences to persistent storage
+     *
+     * @param geofences List of geofences to save
+     */
+    fun storeGeofences(geofences: List<KlaviyoGeofence>) {
```
Comment (belleklaviyo) [inline sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: oooh actually haven't done this on iOS, I think its part of milestone 2? but for parity I could do this too soon/er

Comment (evan-masseau) [reply sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: btw #342 is the PR that adds the API call code in case you missed it

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 6]
```diff
@@ -7,27 +7,55 @@ import android.content.Context
...
  * Coordinator for all geofencing operations
  * - Starts/stops geofence monitoring based on permission state
  * - Fetches geofences from the Klaviyo backend
  * - Adds/removes geofences to/from the system geofencing APIs
  * - Handles geofence transition intents
- * - CHNL-25300 Handle boot receiver events to re-register geofences on device reboot
+ * - TODO CHNL-25300 Handle boot receiver events to re-register geofences on device reboot
  */
-internal object KlaviyoLocationManager : LocationManager {
+internal class KlaviyoLocationManager : LocationManager {
```
[the remark this answers is not in the document]
Comment (dan-peluso) [reply sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: But we do still want this to behave as a singleton right? Ik we have a lot built around the registry to make that happen but just want to verify

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 7]
```diff
@@ -48,6 +69,27 @@ internal object KlaviyoLocationManager : LocationManager {
         )
     }
 
+    /**
+     * Thread-safe list of observers for geofence fetch events
+     */
+    private val observers = Collections.synchronizedList(
```
Comment (dan-peluso) [inline sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: so we are going with the doubly protective approach? I'd be down to cut to one thread safe thing here tbh

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 8]
```diff
@@ -48,6 +69,27 @@ internal object KlaviyoLocationManager : LocationManager {
...
+        observers += callback
+    }
+
+    override fun offGeofenceSync(callback: GeofenceObserver) {
+        observers -= callback
+    }
+
+    private fun notifyObservers(geofences: List<KlaviyoGeofence>) {
+        synchronized(observers) {
+            observers.forEach { it(geofences) }
```
Comment (dan-peluso) [inline sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: could we name the 'it' param to make this a bit more readable?

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 9]
```diff
@@ -73,17 +115,106 @@ internal object KlaviyoLocationManager : LocationManager {
...
+        }
+    }
+
+    /**
+     * Retrieve the list of geofences currently stored in persistent storage
+     *
+     * @return List of geofences, or empty list if none are stored or parsing fails
+     */
+    override fun getStoredGeofences(): List<KlaviyoGeofence> = try {
+        val geofences = Registry.dataStore.fetch(GEOFENCES_STORAGE_KEY)?.let {
```
[the remark this answers is not in the document]
Comment (evan-masseau) [reply sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: There is an upper bound of 20 geofences fetched from this API. 
Shared preferences keeps life simple. There's no explicit limit although in practicality there is an upper bound per key of the system's max string size (1.4MB?). Since we don't need to store fences individually, query them, or separate our the properties, I don't see a compelling need to upgrade our storage engine for this project.

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 10]
```diff
@@ -73,17 +115,106 @@ internal object KlaviyoLocationManager : LocationManager {
         }
     }
 
+    /**
+     * Fetch geofences from the Klaviyo backend using an immediate API call via coroutines.
+     * This bypasses the API queue and makes the request right away.
+     * On success, parses the JSON response into KlaviyoGeofence objects and saves them locally.
+     */
+    fun fetchGeofences() {
+        Registry.log.info("Fetching geofences for company ${Registry.config.apiKey}")
```
Comment (dan-peluso) [inline sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: what level does info show? wondering if this should maybe be a debug log

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 11]
```diff
@@ -73,17 +115,106 @@ internal object KlaviyoLocationManager : LocationManager {
...
+        try {
+            // Serialize geofences to JSON array
+            val jsonArray = JSONArray().apply {
+                geofences.forEach { geofence ->
+                    put(geofence.toJson())
+                }
+            }
+
+            // Store in dataStore
+            Registry.dataStore.store(GEOFENCES_STORAGE_KEY, jsonArray.toString())
```
[the remark this answers is not in the document]
Comment (evan-masseau) [reply sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: replied in other thread

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt (hunk 10 above)
Comment (dan-peluso) [reply sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: Yeah dude it's all over the place with levels (although maybe we should consider giving the claude.md some more context on what levels we want for things more generically)

File: sdk/location/src/main/java/com/klaviyo/location/Geofencing.kt (hunk 1 above)
Comment (ajaysubra) [reply sdk/location/src/main/java/com/klaviyo/location/Geofencing.kt]: lol I'm glad I'm not the only one for whom negation is hard!

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt (hunk 9 above)
[the remark this answers is not in the document]
Comment (ajaysubra) [reply sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: I had the same question, do you foresee a future where we similar to network layer pull out our storage layer from analytics/push? and into it's own foundational module and use that for all things storage? May be this storage layer can leverage any underlying option like writing to disk vs. using shared preference based on the needs of the caller?

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 12]
```diff
@@ -73,17 +115,106 @@ internal object KlaviyoLocationManager : LocationManager {
...
+            Registry.log.verbose("Saved ${geofences.size} geofences to persistent storage")
+        } catch (e: Exception) {
+            Registry.log.error("Failed to save geofences", e)
+        }
+    }
+
+    /**
+     * Remove all geofences from persistent storage
+     */
+    fun clearStoredGeofences() {
```
[the remark this answers is not in the document]
Comment (evan-masseau) [reply sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: Oh if company ID changes we should probably clear... i'll keep it around we may need it later.

File: sdk/location/src/main/java/com/klaviyo/location/LocationManager.kt [hunk 13]
```diff
@@ -0,0 +1,38 @@
+package com.klaviyo.location
+
+import android.content.Context
+import android.content.Intent
+
+typealias GeofenceObserver = (List<KlaviyoGeofence>) -> Unit
+
+interface LocationManager {
```
[the remark this answers is not in the document]
Comment (evan-masseau) [reply sdk/location/src/main/java/com/klaviyo/location/LocationManager.kt]: FYI this isn't apart of the API surface for the geofencing module. I see the confusion, because it is a public interface, but the _implementation_ is inaccessible (except through `core`). The actual API surface is defined in `Klaviyo` extension functions, which I believe we're still aligned between android and iOS right now, `registerGeofencing` and `unregisterGeofencing`

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 14]
```diff
@@ -73,17 +112,104 @@ internal object KlaviyoLocationManager : LocationManager {
         }
     }
 
+    /**
```
[the remark this answers is not in the document]
Comment (evan-masseau) [reply sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: ya, addressing in next PR.

File: sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt [hunk 15]
```diff
@@ -73,17 +114,106 @@ internal object KlaviyoLocationManager : LocationManager {
...
+            }
+
+            // Store in dataStore
+            Registry.dataStore.store(GEOFENCES_STORAGE_KEY, jsonArray.toString())
+            Registry.log.verbose("Saved ${geofences.size} geofences to persistent storage")
+
+            // Notify observers that new geofences have been fetched and stored
+            notifyObservers(geofences)
+        } catch (e: Exception) {
+            Registry.log.error("Failed to save geofences", e)
```
Comment (dan-peluso) [inline sdk/location/src/main/java/com/klaviyo/location/KlaviyoLocationManager.kt]: nit 
```suggestion
            Registry.log.error("Failed to save geofences to local storage", e)
```
